### ISSUE TITLE
Document how to set up ENS for non-mainnet addresses

### ISSUE THREAD

**Issue by carver** [pipermerriam/web3.py#425]
### What was wrong?

No documentation available on using ENS names on non-mainnet. From #342 

### How can it be fixed?

Write docs showing how to set up ENS for other networks.

**Bonus**: Get the chain ID from the client and infer the appropriate ENS registry address.

**Comment by boxydog** [comment]
I am having this same problem.

Below is some code that works on mainnet but not rinkeby.

The [documentation of how to set up geth](https://web3py.readthedocs.io/en/stable/middleware.html#geth-style-proof-of-authority) is not sufficiently detailed to get it to work. The docs describe what I think is a local server, and I don't know how to run one.

```
#!/usr/bin/env python

from web3 import Web3
from web3.auto.infura import (
    INFURA_MAINNET_DOMAIN, build_http_headers, build_infura_url)
from web3.auto.infura.endpoints import INFURA_RINKEBY_DOMAIN
from web3.providers.auto import load_provider_from_uri

from ens import ENS
from eth_typing import ChecksumAddress

_headers = build_http_headers()

# NOTE: build_infura_url requires WEB3_INFURA_PROJECT_ID
# and WEB3_INFURA_API_SECRET environment variables set.
if True:
    # Works with mainnet
    _infura_url = build_infura_url(INFURA_MAINNET_DOMAIN)
else:
    # Doesn't work with Rinkeby
    #
    # "web3.exceptions.ExtraDataLengthError: The field extraData is 97 bytes,
    #  but should be 32. It is quite likely that you are connected to a POA chain.
    #  Refer to http://web3py.readthedocs.io/en/stable/middleware.html#geth-style-proof-of-authority
    #  for more details."
    _infura_url = build_infura_url(INFURA_RINKEBY_DOMAIN)

w3 = Web3(load_provider_from_uri(_infura_url, _headers))
ens = ENS.fromWeb3(w3)

the_ens = ens.name(
    ChecksumAddress('0x5B2063246F2191f18F2675ceDB8b28102e957458'))
print(the_ens)  # jasoncarver.eth
```

**Comment by fselmo** [comment]
Hey @boxydog. The docs could be a bit more clear for other cases but the important piece here is that if you are seeing the extradata error, injecting the `geth_poa_middleware` is the way to resolve it. See the second example in that link for how to import it and inject it into the middleware onion so that you do not see the error you are seeing with the `extraData`.

I'll leave this open for now so update here if that resolves it for you or if you run into any other issues.

**Comment by boxydog** [comment]
This issue hints that there may be an ENS registry somewhere for Rinkeby, but I don't know what that means or how to configure it.

I am saying that a complete and working example documented somewhere would really help, instead of just fragments.

**Comment by boxydog** [comment]
Additionally, if I try

```
from web3.auto.gethdev import w3
print(w3.clientVersion)
```

I just get an error.

```
FileNotFoundError: cannot connect to IPC socket at path: None
```

Pretty sure it's expecting something running and configured.  We want to run our web3.py on Heroku, it's not clear to me how to even run a geth server at all, let alone on Heroku.

cc @fselmo

**Comment by boxydog** [comment]
Can ENS lookups even differ between mainnet and rinkeby? Perhaps I should be doing ENS lookups on mainnet even when our app is using rinkeby for everything else.

**Comment by fselmo** [comment]
Hey @boxydog thanks for the updates. It looks like `ENS.fromWeb3()` injects another middleware so the `geth_poa_middlewre` needs to be injected after that line as well like this:

``` python
ens = ENS.fromWeb3(w3)
ens.web3.middleware_onion.inject(geth_poa_middleware, layer=0)

the_ens = ens.name(
    ChecksumAddress('0x5B2063246F2191f18F2675ceDB8b28102e957458')
)
```

This should get you unstuck for now but we can definitely use a documentation improvement around some of this so I will leave this issue open and use this as a guide to make some updates.

Thanks again.

**Comment by boxydog** [comment]
Thanks very much for your response. I still get the same error.

Here is all the code after the `if` just to double-check:

```
w3 = Web3(load_provider_from_uri(_infura_url, _headers))

ens = ENS.fromWeb3(w3)

from web3.middleware import geth_poa_middleware
w3.middleware_onion.inject(geth_poa_middleware, layer=0)

the_ens = ens.name(
    ChecksumAddress('0x5B2063246F2191f18F2675ceDB8b28102e957458'))
print(the_ens)  # jasoncarver.eth
```

**Comment by boxydog** [comment]
Also, here is my version:

```
$ python -c "import web3; print(web3.__version__)"
5.25.0
```

**Comment by fselmo** [comment]
@boxydog yes please copy exactly what I wrote. You need `ens.web3.middleware_onion.inject(....)` (or, alternatively, inject the middleware into w3 before instantiating `ENS.from_web3(w3)`).

**Comment by boxydog** [comment]
My apologies, I didn't see that difference.

Yes, now it doesn't throw an error. It returns `None`, but that may well be correct. Perhaps that address is not mapped in Rinkeby ENS space, whatever that means. :)

Thanks for your help!
